lib/client/conntest/database.go

```diff
@@ -0,0 +1,408 @@
+	// UserClient is an auth client that has a User's identity.
+	UserClient ClientDatabaseConnectionTester
+
+	// PublicProxyAddr is public address of the proxy
+	PublicProxyAddr string
+
+	// TLSRoutingEnabled indicates that proxy supports ALPN SNI server where
+	// all proxy services are exposed on a single TLS listener (Proxy Web Listener).
+	TLSRoutingEnabled bool
+}
+
+// DatabaseConnectionTester implements the ConnectionTester interface for Testing Database access.
+type DatabaseConnectionTester struct {
+	cfg DatabaseConnectionTesterConfig
+}
+
+// NewDatabaseConnectionTester returns a new DatabaseConnectionTester
+func NewDatabaseConnectionTester(cfg DatabaseConnectionTesterConfig) (*DatabaseConnectionTester, error) {
+	return &DatabaseConnectionTester{
+		cfg: cfg,
+	}, nil
+}
+
+// TestConnection tests the access to a database using:
+// - auth Client using the User access
+// - the resource name
+// - database user and database name to connect to
+//
+// A new ConnectionDiagnostic is created and used to store the traces as it goes through the checkpoints
+// To connect to the Database, we will create a cert-key pair and setup a Database client back to Teleport Proxy.
+// The following checkpoints are reported:
+// - database server for the requested database exists / the user's roles can access it
+// - the user can use the requested database user and database name (per their roles)
+// - the database is acessible and accepting connections from the database server
+// - the database has the database user and database name that was requested
+func (s *DatabaseConnectionTester) TestConnection(ctx context.Context, req TestConnectionRequest) (types.ConnectionDiagnostic, error) {
+	if req.ResourceKind != types.KindDatabase {
+		return nil, trace.BadParameter("invalid value for ResourceKind, expected %q got %q", types.KindDatabase, req.ResourceKind)
+	}
+
+	connectionDiagnosticID := uuid.NewString()
+	connectionDiagnostic, err := types.NewConnectionDiagnosticV1(
+		connectionDiagnosticID,
+		map[string]string{},
+		types.ConnectionDiagnosticSpecV1{
+			// We start with a failed state so that we don't need to set it to each return statement once an error is returned.
+			// if the test reaches the end, we force the test to be a success by calling
+			// 	connectionDiagnostic.SetMessage(types.DiagnosticMessageSuccess)
+			//	connectionDiagnostic.SetSuccess(true)
+			Message: types.DiagnosticMessageFailed,
+		},
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := s.cfg.UserClient.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	databaseServers, err := s.getDatabaseServers(ctx, req.ResourceName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(databaseServers) == 0 {
+		connDiag, err := s.appendDiagnosticTrace(ctx,
+			connectionDiagnosticID,
+			types.ConnectionDiagnosticTrace_RBAC_DATABASE,
+			"You are not authorized to access this Database. "+
```

Comment:
I think we should lead with saying that database wasn't found and then explain that this could happen either because of authz issues or database agent matching this database isn't running.